src/scatter.h

```diff
@@ -33,222 +35,335 @@ template <class T = float> class FormFactorSphere {
  */
-template <class Tformfactor, class Tgeometry = Geometry::Chameleon, class T = float> class DebyeFormula {
-  protected:
-    Tformfactor F;    // scattering from a single particle
-    Tgeometry geo;    // geometry to use for distance calculations
-    std::map<T, T> I; //!< Sampled, average I(q)
-    std::map<T, T> S; //!< Weighted number of samplings
-    T rc;
-
-  public:
-    T qmin, qmax, dq;
-
-    DebyeFormula(const json &j) {
-        geo = R"( { "type": "sphere", "radius": 1e9 } )"_json;
-        dq = j.at("dq").get<double>();
-        qmin = j.at("qmin").get<double>();
-        qmax = j.at("qmax").get<double>();
-        rc = j.value("cutoff", 1.0e9);
+template <class Tformfactor, class T = float> class DebyeFormula {
+    static constexpr T r_cutoff_infty = 1e9; //<! a cutoff distance in angstrom considered to be infinity
+    T q_mesh_min, q_mesh_max, q_mesh_step; //<! q_mesh parameters in inverse angstrom; used for inline lambda-functions
 
-        if (dq <= 0 || qmin <= 0 || qmax <= 0 || qmin > qmax)
-            throw std::runtime_error("DebyeFormula: invalid q parameters");
+    /**
+     * @param m mesh point index
+     * @return the scattering vector magnitude q at the mesh point m
+     */
+    #pragma omp declare simd uniform(this) linear(m:1)
+    T q_mesh(int m) {
+        return q_mesh_min + m * q_mesh_step;
     }
 
     /**
-     * @brief Sample I(q) and add to average
-     *
-     * The q range is read from input as `qmin`, `qmax`, `dq` in units of
-     * inverse angstrom.
-     * It is also possible to specify an isotropic correction beyond
-     * a given cut-off -- see for example
-     * <https://debyer.readthedocs.org/en/latest/>.
-     * The cut-off distance - which should be smaller than half the
-     * box length for a cubic system is read specific by the
-     * keyword `sofq_cutoff`.
+     * @brief Initialize mesh for intensity and sampling.
+     * @param q_min Minimum q-value to sample (1/A)
+     * @param q_max Maximum q-value to sample (1/A)
+     * @param q_step Spacing between mesh points (1/A)
      */
-    template <class Tpvec> void sample(const Tpvec &p, T f = 1, T V = -1) {
-        assert(qmin > 0 && qmax > 0 && dq > 0 && "q range invalid.");
-        sample(p, qmin, qmax, dq, f, V);
+    void init_mesh(T q_min, T q_max, T q_step) {
+        if (q_step <= 0 || q_min <= 0 || q_max <= 0 || q_min > q_max ||
+            q_step / q_max < 4 * std::numeric_limits<T>::epsilon()) {
+            throw std::range_error("DebyeFormula: Invalid mesh parameters for q");
+        }
+        q_mesh_min = q_min < 1e-6 ? q_step : q_min; // ensure that q > 0
+        q_mesh_max = q_max;
+        q_mesh_step = q_step;
+        // resolution of the 1D mesh approximation of the scattering vector magnitude q
+        const size_t q_resolution = 1 + (size_t)((q_max - q_min) / q_step);
+        if (q_resolution > std::numeric_limits<int>::max()) { // we use int index variable
```

Comment:
if overstepping, unsigned int becomes zero. can this condition evaluate to true?

Reply:
Let’s do it properly. Inspired by boost, I add `numeric_cast` function into `auxiliary.h`. Here I cast and throw an exception.